File: ch03/14/reOrderArray.cpp
class Solution {
public:
    void reOrderArray(vector<int> &array) {
        int len = array.size();
        if (0 == len || 1 == len) 
            return;
        int point1 = 0, point2;
        while (point1 < len) {
            while (array[point1] % 2 == 1) //point1指向第一个偶数
                point1 ++;
            point2 = point1 + 1;
            if (point2 == len)
                break;
            
            while (array[point2] %2 == 0){ //point2 指向point1之后的第一个奇数
                point2 ++;
                if (point2 == len)
                    return;
            }
            int j = point2;
            int point2_val = array[point2];
            while (j > point1) {
                array[j] = array[j - 1];
            	j--;
            }
            array[point1] = point2_val;
            point1 ++;
        }
    }
};

File: ch03/14/reOrderArray.py
# -*- coding:utf-8 -*-
'''
调整数组顺序使奇数位于偶数前面
'''
class Solution:
    def reOrderArray(self, array):
        # write code here
        array_len = len(array)
        if array_len == 0 or array_len == 1:
            return array

        odd_num = 0
        for num in array:
            if num & 0x1 == 1: 
                odd_num += 1

        odd_point = 0
        while odd_point < odd_num:
            if array[odd_point] & 0x1 == 1:
                odd_point += 1
            else:
                next_point = odd_point
                tmp1 = array[next_point]
                while tmp1 & 0x1 != 1 and next_point + 1 < array_len: 
                    tmp2 = array[next_point + 1]
                    array[next_point + 1] = tmp1
                    tmp1 = tmp2
                    next_point += 1
                array[odd_point] = tmp1
                odd_point += 1

        return array    

if __name__ == "__main__":             
    s  = Solution()
    lst = [1,2,3,4,5,6,7]
    print s.reOrderArray(lst)